rubix-spi/src/main/java/com/qubole/rubix/spi/ClusterManager.java

```diff
@@ -87,17 +84,39 @@ protected String getCurrentNodeHostAddress()
       // Empty result set => server up and only master node running, return localhost has the only node
       // Do not need to consider failed nodes list as 1node cluster and server is up since it replied to allNodesRequest
       nodes = ImmutableList.of(getCurrentNodeHostAddress());
-    } else {
-      Collections.sort(nodes);
     }
 
-    currentNodeIndex = nodes.indexOf(getCurrentNodeHostname());
-    if (currentNodeIndex == -1) {
-      currentNodeIndex = nodes.indexOf(getCurrentNodeHostAddress());
+    // remove stale nodes from consistent hash ring
+    for (SimpleNode ringNode : consistentHashRing.getNodes()) {
+      if (!nodes.contains(ringNode.getKey()))
+      {
+        log.debug("Removing node: " + ringNode.getKey() + " from consistent hash ring");
+        consistentHashRing.remove(ringNode);
+      }
     }
-    if (currentNodeIndex == -1) {
-      log.error(String.format("Could not initialize cluster nodes=%s nodeHostName=%s nodeHostAddress=%s " +
-              "currentNodeIndex=%d", nodes, getCurrentNodeHostname(), getCurrentNodeHostAddress(), currentNodeIndex));
+
+    // add new nodes to consistent hash ring
+    for (String node : nodes) {
+      SimpleNode ringNode = SimpleNode.of(node);
+      if (!consistentHashRing.contains(ringNode)) {
+        log.debug("Adding node: " + ringNode.getKey() + " to consistent hash ring");
+        consistentHashRing.add(ringNode);
+      }
+    }
+
+    log.debug("Total nodes in consistent hash ring: " + consistentHashRing.getNodes());
+
+    if (currentNodeName.isEmpty()) {
+      if (consistentHashRing.contains(SimpleNode.of(nodeHostname))) {
```

Comment:
This has bug, change this to:
```
if (currentNodeName.isEmpty()) {
      if (consistentHashRing.contains(SimpleNode.of(getCurrentNodeHostname()))) {
        currentNodeName = getCurrentNodeHostname();
      }
      else if (consistentHashRing.contains(SimpleNode.of(getCurrentNodeHostAddress()))) {
        currentNodeName = getCurrentNodeHostAddress();
      }
      else {
        log.error(String.format("Could not initialize cluster nodes=%s nodeHostName=%s nodeHostAddress=%s " +
                "currentNodeIndex=%s", nodes, getCurrentNodeHostname(), getCurrentNodeHostAddress(), currentNodeName));
      }
    }if (currentNodeName.isEmpty()) {
      if (consistentHashRing.contains(SimpleNode.of(getCurrentNodeHostname()))) {
        currentNodeName = getCurrentNodeHostname();
      }
      else if (consistentHashRing.contains(SimpleNode.of(getCurrentNodeHostAddress()))) {
        currentNodeName = getCurrentNodeHostAddress();
      }
      else {
        log.error(String.format("Could not initialize cluster nodes=%s nodeHostName=%s nodeHostAddress=%s " +
                "currentNodeIndex=%s", nodes, getCurrentNodeHostname(), getCurrentNodeHostAddress(), currentNodeName));
      }
    }
```

Reply:
Sorry, I don't understand the bug part. I am accessing the class variables directly instead of calling getters in condition.

Reply:
don't use `nodeHostname` and `nodeAddress` directly, instead use `getCurrentNodeHostname` and `getCurrentNodeHostAddress` so that any overridden values by specific implementations is used.

Reply:
fixed.